In [ ]:
import numpy as np
import pandas as pd
import io
from google.colab import files

In [ ]:
uploaded = files.upload()
patients = pd.read_csv(io.StringIO(uploaded['patients.csv'].decode('utf-8')))
patients

Saving patients.csv to patients (2).csv


patient_id assigned_sex given_name  ... weight height   bmi
0             1       female        Zoe  ...  121.7     66  19.6
1             2       female     Pamela  ...  118.8     66  19.2
2             3         male        Jae  ...  177.8     71  24.8
3             4         male       Liêm  ...  220.9     70  31.7
4             5         male        Tim  ...  192.3     27  26.1
..          ...          ...        ...  ...    ...    ...   ...
498         499         male    Mustafa  ...  181.1     72  24.6
499         500         male      Ruman  ...  239.6     70  34.4
500         501       female      Jinke  ...  171.2     67  26.8
501         502       female    Chidalu  ...  176.9     67  27.7
502         503         male        Pat  ...  138.2     71  19.3

[503 rows x 14 columns]

In [ ]:
uploaded = files.upload()
treatment = pd.read_csv(io.StringIO(uploaded['treatments.csv'].decode('utf-8')))
treatment

Saving treatments.csv to treatments (2).csv


given_name       surname    auralin  ... hba1c_start  hba1c_end  hba1c_change
0     veronika      jindrová  41u - 48u  ...        7.63       7.20           NaN
1       elliot    richardson          -  ...        7.56       7.09          0.97
2     yukitaka      takenaka          -  ...        7.68       7.25           NaN
3         skye    gormanston  33u - 36u  ...        7.97       7.62          0.35
4       alissa        montez          -  ...        7.78       7.46          0.32
..         ...           ...        ...  ...         ...        ...           ...
275     albina      zetticci  45u - 51u  ...        7.93       7.73          0.20
276       john   teichelmann          -  ...        7.90       7.58           NaN
277     mathea       lillebø  23u - 36u  ...        9.04       8.67          0.37
278     vallie        prince  31u - 38u  ...        7.64       7.28          0.36
279     samúel  guðbrandsson  53u - 56u  ...        8.00       7.64          0.36

[280 rows x 7 columns]

In [ ]:
uploaded = files.upload()
adverse = pd.read_csv(io.StringIO(uploaded['adverse_reactions.csv'].decode('utf-8')))
adverse

Saving adverse_reactions.csv to adverse_reactions (2).csv


given_name     surname           adverse_reaction
0         berta  napolitani  injection site discomfort
1          lena        baer               hypoglycemia
2        joseph         day               hypoglycemia
3        flavia  fiorentino                      cough
4       manouck     wubbels          throat irritation
5       jasmine       sykes               hypoglycemia
6        louise     johnson               hypoglycemia
7       albinca     komavec               hypoglycemia
8           noe      aranda               hypoglycemia
9         sofia   hermansen  injection site discomfort
10        tegan     johnson                   headache
11         abel     yonatan                      cough
12    abdul-nur         isa               hypoglycemia
13         leon      scholz  injection site discomfort
14     gabriele     saenger               hypoglycemia
15       jia li        teng                     nausea
16        jakob    jakobsen               hypoglycemia
17  christopher    woodward                     nausea
18          ole    petersen               hypoglycemia
19       finley    chandler                   headache
20     anenechi       chidi               hypoglycemia
21     miłosław  wiśniewski  injection site discomfort
22        lixue       hsueh  injection site discomfort
23        merci      leroux               hypoglycemia
24         kang         mai  injection site discomfort
25       elliot  richardson               hypoglycemia
26      clinton      miller          throat irritation
27       idalia       moore               hypoglycemia
28       xiuxiu       chang               hypoglycemia
29         alex    crawford               hypoglycemia
30       monika      lončar               hypoglycemia
31       steven         roy                   headache
32      cecilie      nilsen               hypoglycemia
33    krisztina      magyar               hypoglycemia

In [ ]:
patients.shape

(503, 14)

In [ ]:
treatment.shape

(280, 7)

In [ ]:
adverse.shape

(34, 3)

In [ ]:
patients.sample(5)

patient_id assigned_sex given_name  ... weight height   bmi
489         490       female    Jasmine  ...  187.2     63  33.2
31           32         male      Tomáš  ...  133.3     70  19.1
255         256       female      Mette  ...  187.1     60  36.5
427         428       female      Tecla  ...  174.7     61  33.0
30           31       female     Janina  ...  156.9     65  26.1

[5 rows x 14 columns]

In [ ]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   patient_id    503 non-null    int64  
 1   assigned_sex  503 non-null    object 
 2   given_name    503 non-null    object 
 3   surname       503 non-null    object 
 4   address       491 non-null    object 
 5   city          491 non-null    object 
 6   state         491 non-null    object 
 7   zip_code      491 non-null    float64
 8   country       491 non-null    object 
 9   contact       491 non-null    object 
 10  birthdate     503 non-null    object 
 11  weight        503 non-null    float64
 12  height        503 non-null    int64  
 13  bmi           503 non-null    float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


In [ ]:
patients[patients['address'].duplicated()]

patient_id assigned_sex given_name  ... weight height   bmi
29           30         male       Jake  ...  155.8     67  24.4
219         220         male         Mỹ  ...  237.8     69  35.1
229         230         male       John  ...  180.0     72  24.4
230         231       female  Elisabeth  ...  165.9     63  29.4
234         235       female    Martina  ...  199.5     65  33.2
237         238         male       John  ...  180.0     72  24.4
242         243         male       John  ...  205.3     74  26.4
244         245         male       John  ...  180.0     72  24.4
249         250         male   Benjamin  ...  146.5     69  21.6
251         252         male       John  ...  180.0     72  24.4
257         258         male        Jin  ...  231.7     69  34.2
264         265       female   Wafiyyah  ...  158.6     63  28.1
269         270       female     Flavia  ...  175.2     61  33.1
277         278         male       John  ...  180.0     72  24.4
278         279       female   Generosa  ...  124.3     69  18.4
282         283       female      Sandy  ...  206.1     64  35.4
286         287         male      Lewis  ...  155.3     68  23.6
296         297       female        Chỉ  ...  181.1     63  32.1
502         503         male        Pat  ...  138.2     71  19.3

[19 rows x 14 columns]

In [ ]:
treatment.sample(10)

given_name      surname    auralin  ... hba1c_start  hba1c_end  hba1c_change
130     kamila     pecinová          -  ...        7.77       7.39          0.38
237    manouck      wubbels  55u - 62u  ...        7.66       7.40          0.26
110      suoma     koivunen          -  ...        7.59       7.13           NaN
85      nilton  quintanilla  42u - 49u  ...        7.86       7.48          0.38
95      jaakko    eskelinen  39u - 45u  ...        7.73       7.42          0.31
137      ivona       jakšić          -  ...        7.98       7.54           NaN
232     pamela         hill          -  ...        9.53       9.10          0.93
161    rebecca       priest  40u - 42u  ...        7.87       7.57          0.30
172     malina      balagić          -  ...        7.65       7.30           NaN
267       leon       scholz          -  ...        7.72       7.29          0.93

[10 rows x 7 columns]

In [ ]:
treatment.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    280 non-null    object 
 1   surname       280 non-null    object 
 2   auralin       280 non-null    object 
 3   novodra       280 non-null    object 
 4   hba1c_start   280 non-null    float64
 5   hba1c_end     280 non-null    float64
 6   hba1c_change  171 non-null    float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


In [ ]:
treatment[treatment.duplicated()]

given_name surname    auralin novodra  hba1c_start  hba1c_end  hba1c_change
136     joseph     day  29u - 36u       -          7.7       7.19           NaN

In [ ]:
adverse.sample(5)

given_name   surname   adverse_reaction
14   gabriele   saenger       hypoglycemia
15     jia li      teng             nausea
6      louise   johnson       hypoglycemia
18        ole  petersen       hypoglycemia
4     manouck   wubbels  throat irritation

In [ ]:
adverse.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 3 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   given_name        34 non-null     object
 1   surname           34 non-null     object
 2   adverse_reaction  34 non-null     object
dtypes: object(3)
memory usage: 944.0+ bytes


In [ ]:
patients.describe()

patient_id      zip_code      weight      height         bmi
count  503.000000    491.000000  503.000000  503.000000  503.000000
mean   252.000000  49084.118126  173.434990   66.634195   27.483897
std    145.347859  30265.807442   33.916741    4.411297    5.276438
min      1.000000   1002.000000   48.800000   27.000000   17.100000
25%    126.500000  21920.500000  149.300000   63.000000   23.300000
50%    252.000000  48057.000000  175.300000   67.000000   27.200000
75%    377.500000  75679.000000  199.500000   70.000000   31.750000
max    503.000000  99701.000000  255.900000   79.000000   37.700000

In [ ]:
treatment.describe()

hba1c_start   hba1c_end  hba1c_change
count   280.000000  280.000000    171.000000
mean      7.985929    7.589286      0.546023
std       0.568638    0.569672      0.279555
min       7.500000    7.010000      0.200000
25%       7.660000    7.270000      0.340000
50%       7.800000    7.420000      0.380000
75%       7.970000    7.570000      0.920000
max       9.950000    9.580000      0.990000

# **Quality**

[Completeness issue(C), Validity issue(V), Accurate issue(A), Consistent issue(IC)] 

**Patients**

1. Certain zip codes have 4 digits.(V)
2. Incorrect data types for assigned_sex, zip_code, birthdate.(V)
3. For the state column sometimes full names are used and other times short forms are used.(IC)
4. Contact details of 12 patients are missing.(C)
5. Inconsistent way of storing phone numbers in contact column.(IC)
6. Errorness data for weights and heights.(A)
7. There are 19 duplicates in patients dataframe.(IC)
8. Data types of patient_id, weight and bmi can be changed to reduce the memory storage.(A)

 **Treatments**
1. hba1c_change has NaN values.(C)
2. given_name and surname has first letter in lowercase.(IC)
3. auralin and novodra has u attached with numbers.(V)
4. Only 280 rows are there instead of 350.(C)
5. Use of dashes in auralin and novodra column instead of NaN elements.(IC)
6. Incorrect data type for auralin and novadra columns.(V)
7. Data types of hba1c_start, hba1c_end, hba1c_change can be changed to 32-bit to reduce the memory storage.(A)
8. There is 1 duplicate in treatment dataframe.(IC)
9. hba1c_change has 9 instead of 4.(A)

**Adverse Reactions**
1. given_name and surname has first letter in lowercase.(IC)

# **Tidiness**

**Patients**
1. Phone number and email in the same column i.e. the contact column.

**Treatments**
1. Three variables(informations) stored in auralin and novodra columns.

**Adverse Reactions**
1. No point of using adverse_reaction column.

# **Cleaning Data**

In [ ]:
# Creating copies of the dataframes

patients_copy=patients.copy()
treatment_copy=treatment.copy()
adverse_copy=adverse.copy()

**1. Completeness Issues**
        
        a. Contact details of 12 patients are missing. [Done]
        b. hba1c_change has NaN values.[Done]
        c. Only 280 rows are there instead of 350. [Done]

In [ ]:
patients_copy[patients_copy['address'].isnull()]

patient_id assigned_sex given_name  ... weight height   bmi
209         210       female     Lalita  ...  143.4     62  26.2
219         220         male         Mỹ  ...  237.8     69  35.1
230         231       female  Elisabeth  ...  165.9     63  29.4
234         235       female    Martina  ...  199.5     65  33.2
242         243         male       John  ...  205.3     74  26.4
249         250         male   Benjamin  ...  146.5     69  21.6
257         258         male        Jin  ...  231.7     69  34.2
264         265       female   Wafiyyah  ...  158.6     63  28.1
269         270       female     Flavia  ...  175.2     61  33.1
278         279       female   Generosa  ...  124.3     69  18.4
286         287         male      Lewis  ...  155.3     68  23.6
296         297       female        Chỉ  ...  181.1     63  32.1

[12 rows x 14 columns]

**Define**

70 rows are missing in treatment dataframe so we will concatenate it with treatments_cut dataframe.

In [ ]:
uploaded = files.upload()
treatments_cut = pd.read_csv(io.StringIO(uploaded['treatments_cut.csv'].decode('utf-8')))
treatments_cut

Saving treatments_cut.csv to treatments_cut (2).csv


given_name     surname    auralin  ... hba1c_start  hba1c_end  hba1c_change
0       jožka   resanovič  22u - 30u  ...        7.56       7.22          0.34
1   inunnguaq    heilmann  57u - 67u  ...        7.85       7.45           NaN
2       alwin    svensson  36u - 39u  ...        7.78       7.34           NaN
3         thể       lương          -  ...        7.64       7.22          0.92
4      amanda     ribeiro  36u - 44u  ...        7.85       7.47          0.38
..        ...         ...        ...  ...         ...        ...           ...
65     rovzan     kishiev  32u - 37u  ...        7.75       7.41          0.34
66      jakob    jakobsen          -  ...        7.96       7.51          0.95
67      bernd   schneider  48u - 56u  ...        7.74       7.44          0.30
68      berta  napolitani          -  ...        7.68       7.21           NaN
69     armina       sauvé  36u - 46u  ...        7.86       7.40           NaN

[70 rows x 7 columns]

**Code**

In [ ]:
frames = [treatment_copy, treatments_cut]
treatment_copy = pd.concat(frames)

**Test**

In [ ]:
treatment_copy.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 350 entries, 0 to 69
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    350 non-null    object 
 1   surname       350 non-null    object 
 2   auralin       350 non-null    object 
 3   novodra       350 non-null    object 
 4   hba1c_start   350 non-null    float64
 5   hba1c_end     350 non-null    float64
 6   hba1c_change  213 non-null    float64
dtypes: float64(3), object(4)
memory usage: 21.9+ KB


**Define**

Subtract hba1c_end from hba1c_start to get hba1c_change.

**Code**

In [ ]:
treatment_copy['hba1c_change']=treatment_copy['hba1c_start']-treatment_copy['hba1c_end']

**Test**

In [ ]:
treatment_copy.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 350 entries, 0 to 69
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    350 non-null    object 
 1   surname       350 non-null    object 
 2   auralin       350 non-null    object 
 3   novodra       350 non-null    object 
 4   hba1c_start   350 non-null    float64
 5   hba1c_end     350 non-null    float64
 6   hba1c_change  350 non-null    float64
dtypes: float64(3), object(4)
memory usage: 21.9+ KB


# **2. Tidiness Issue**
     a. Phone number and email in the same column i.e. the contact column.
     b. Three variables(informations) stored in auralin and novodra columns.
     c. No point of using adverse_reaction column.

**Define**

Making separate columns for phone numbers and email and dropping the contact column and keeping the formats of the phone number as it is, because US has various formats for local,domestic,international etc.

**Code**

In [ ]:
patients_copy['phone_number']=patients_copy.contact.str.extract('((?:\+\d{1,2}\s)?\(?\d{3}\)?[\s.-]?\d{3}[\s.-]?\d{4})', expand=True)

In [ ]:
patients_copy['email']=patients_copy.contact.str.extract('([a-zA-Z][a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+[a-zA-Z])', expand=True)

In [ ]:
patients_copy = patients_copy.drop('contact', axis=1)

**Test**

In [ ]:
patients_copy.head()

patient_id assigned_sex  ...       phone_number                      email
0           1       female  ...       951-719-9170   ZoeWellish@superrito.com
1           2       female  ...  +1 (217) 569-3204       PamelaSHill@cuvox.de
2           3         male  ...       402-363-6804       JaeMDebord@gustr.com
3           4         male  ...  +1 (732) 636-8246  PhanBaLiem@jourrapide.com
4           5         male  ...       334-515-7487        TimNeudorf@cuvox.de

[5 rows x 15 columns]

**Define**

Creating one column for auralin and novodra by the melt function and creating two separate column for the start and end dose.

**Code**

In [ ]:
treatment_copy=pd.melt(treatment_copy, id_vars=['given_name', 'surname', 'hba1c_start', 'hba1c_end', 'hba1c_change'],var_name='treatment', value_name='dose')

In [ ]:
treatment_copy=treatment_copy[treatment_copy['dose']!='-']

In [ ]:
treatment_copy['start'],treatment_copy['end']=treatment_copy['dose'].str.split('-').str

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:1: FutureWarning: Columnar iteration over characters will be deprecated in future releases.
  """Entry point for launching an IPython kernel.


In [ ]:
treatment_copy.drop(columns={'dose'}, inplace=True)

**Test**

In [ ]:
treatment_copy.tail(5)

given_name     surname  hba1c_start  ...  treatment  start   end
688  christopher    woodward         7.51  ...    novodra   55u    51u
690        maret    sultygov         7.67  ...    novodra   26u    23u
694        lixue       hsueh         9.21  ...    novodra   22u    23u
696        jakob    jakobsen         7.96  ...    novodra   28u    26u
698        berta  napolitani         7.68  ...    novodra   42u    44u

[5 rows x 8 columns]

In [ ]:
treatment_copy.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 350 entries, 0 to 698
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    350 non-null    object 
 1   surname       350 non-null    object 
 2   hba1c_start   350 non-null    float64
 3   hba1c_end     350 non-null    float64
 4   hba1c_change  350 non-null    float64
 5   treatment     350 non-null    object 
 6   start         350 non-null    object 
 7   end           350 non-null    object 
dtypes: float64(3), object(5)
memory usage: 24.6+ KB


**Define**

Merging the adverse_reaction column in treatment dataframe.

**Code**

In [ ]:
treatment_copy=pd.merge(treatment_copy, adverse_copy,on=['given_name', 'surname'],how='left')

**Test**

In [ ]:
treatment_copy.head()

given_name     surname  hba1c_start  ...  start   end adverse_reaction
0   veronika    jindrová         7.63  ...   41u    48u              NaN
1       skye  gormanston         7.97  ...   33u    36u              NaN
2     sophia      haugen         7.65  ...   37u    42u              NaN
3      eddie      archer         7.89  ...   31u    38u              NaN
4       asia     woźniak         7.76  ...   30u    36u              NaN

[5 rows x 9 columns]

# **3.Quality Issues**
        a. Certain zip codes have 4 digits.(V)
        b. Incorrect data types for assigned_sex, zip_code, birthdate.(V)
        c. For the state column sometimes full names are used and other times short forms are used.(IC)
        d. Errorness data for weights and heights.(A)
        e. given_name and surname has first letter in lowercase.(IC)
        f. There is 1 duplicate in treatment dataframe.(IC)
        g. There are certain John Doe records in patients_copy dataframe.

**Define**

Converting the zip code from float to string using astype
Removing .0 from the zip code and making the zip codes to 5-digit

**Code**

In [ ]:
patients_copy['zip_code']=patients_copy['zip_code'].astype('str').str[:-2].str.pad(5, fillchar='0')

In [ ]:
patients_copy['zip_code']=patients_copy['zip_code'].replace('0000n',np.nan)

**Test**

In [ ]:
patients_copy.sample(10)

patient_id assigned_sex  ...       phone_number                           email
276         277         male  ...  +1 (208) 264-9063         AlexCrawford@dayrep.com
430         431         male  ...       701-662-1983       TimothyJCotton@dayrep.com
118         119         male  ...       337-685-4885    AdibMutazzGhanem@fleckens.hu
32           33       female  ...       562-985-4582       PhamThiBichMot@dayrep.com
158         159         male  ...  +1 (918) 971-5864         ChibuzoOkoli@einrot.com
353         354         male  ...       478-676-7058         OlePetersen@armyspy.com
480         481         male  ...       210 326 5509  NasserMazinMansour@fleckens.hu
435         436       female  ...  +1 (757) 736-7026          IvanaKasna@teleworm.us
383         384         male  ...       208 657 2473     TorbenMMortensen@einrot.com
7             8       female  ...       408 778 3236          XiuxiuChang@einrot.com

[10 rows x 15 columns]

In [ ]:
patients_copy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 15 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   patient_id    503 non-null    int64  
 1   assigned_sex  503 non-null    object 
 2   given_name    503 non-null    object 
 3   surname       503 non-null    object 
 4   address       491 non-null    object 
 5   city          491 non-null    object 
 6   state         491 non-null    object 
 7   zip_code      491 non-null    object 
 8   country       491 non-null    object 
 9   birthdate     503 non-null    object 
 10  weight        503 non-null    float64
 11  height        503 non-null    int64  
 12  bmi           503 non-null    float64
 13  phone_number  491 non-null    object 
 14  email         491 non-null    object 
dtypes: float64(2), int64(2), object(11)
memory usage: 59.1+ KB


**Define**

The incorrect datatypes of assigned_sex and birthdate has been corrected.

**Code**

In [ ]:
patients_copy['assigned_sex']=patients_copy['assigned_sex'].astype('category')

In [ ]:
patients_copy['birthdate']=pd.to_datetime(patients_copy['birthdate'])

**Test**

In [ ]:
patients_copy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 15 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   patient_id    503 non-null    int64         
 1   assigned_sex  503 non-null    category      
 2   given_name    503 non-null    object        
 3   surname       503 non-null    object        
 4   address       491 non-null    object        
 5   city          491 non-null    object        
 6   state         491 non-null    object        
 7   zip_code      491 non-null    object        
 8   country       491 non-null    object        
 9   birthdate     503 non-null    datetime64[ns]
 10  weight        503 non-null    float64       
 11  height        503 non-null    int64         
 12  bmi           503 non-null    float64       
 13  phone_number  491 non-null    object        
 14  email         491 non-null    object        
dtypes: category(1), datetime64[ns](1), float

**Define**

Certain states which are in full form can be converted to shortforms by applying function.

**Code**

In [ ]:
patients_copy['state'].value_counts()

California    36
TX            32
New York      25
CA            24
NY            22
MA            22
PA            18
GA            15
Illinois      14
OH            14
Florida       13
MI            13
LA            13
OK            13
NJ            12
VA            11
IL            10
WI            10
MS            10
TN             9
MN             9
IN             9
AL             9
FL             9
WA             8
KY             8
NC             8
MO             7
NV             6
KS             6
ID             6
SC             5
IA             5
CT             5
ME             4
Nebraska       4
CO             4
AZ             4
AR             4
RI             4
ND             4
WV             3
MD             3
OR             3
SD             3
DE             3
VT             2
NE             2
DC             2
MT             2
NH             1
WY             1
NM             1
AK             1
Name: state, dtype: int64

In [ ]:
state_short={'California':'CA','New York':'NY','Illinois':'IL','Florida':'FL','Nebraska':'NE'}

In [ ]:
def state_abbv(patient):
    if patient['state'] in state_short.keys():
        abbv = state_short[patient['state']]
        return abbv
    else:
        return patient['state']

In [ ]:
patients_copy['state']=patients_copy.apply(state_abbv, axis=1)

**Test**

In [ ]:
patients_copy.sample(10)

patient_id  ...                            email
243         244  ...         MustafaAdonay@dayrep.com
66           67  ...  KonstantynKowalczyk@armyspy.com
158         159  ...          ChibuzoOkoli@einrot.com
361         362  ...      HansPeterRadislav@gustr.com
398         399  ...           IlijaHorvat@dayrep.com
254         255  ...           KangMai@jourrapide.com
232         233  ...          KyoukoOno@superrito.com
285         286  ...             AydenAllan@rhyta.com
247         248  ...     TuukkaLeppaluoto@teleworm.us
354         355  ...          VivianRHouse@dayrep.com

[10 rows x 15 columns]

In [ ]:
patients_copy.describe()

patient_id      weight      height         bmi
count  503.000000  503.000000  503.000000  503.000000
mean   252.000000  173.434990   66.634195   27.483897
std    145.347859   33.916741    4.411297    5.276438
min      1.000000   48.800000   27.000000   17.100000
25%    126.500000  149.300000   63.000000   23.300000
50%    252.000000  175.300000   67.000000   27.200000
75%    377.500000  199.500000   70.000000   31.750000
max    503.000000  255.900000   79.000000   37.700000

**Define**

Correcting the outliers of weight by using formula and loc function.

**Code**

In [ ]:
patients_copy['weight'].sort_values().head()

210     48.8
459    102.1
335    102.7
74     103.2
317    106.0
Name: weight, dtype: float64

In [ ]:
patients_copy[patients_copy.weight == 48.8]

patient_id assigned_sex  ...  phone_number                          email
210         211       female  ...  330-202-2145  CamillaZaitseva@superrito.com

[1 rows x 15 columns]

In [ ]:
weight_pounds=patients_copy[patients_copy.patient_id==211].weight*2.20462

In [ ]:
weight_pounds.values

array([107.585456])

In [ ]:
patients_copy.loc[210, 'weight'] = weight_pounds.values

In [ ]:
patients_copy[patients_copy.patient_id==211]

patient_id assigned_sex  ...  phone_number                          email
210         211       female  ...  330-202-2145  CamillaZaitseva@superrito.com

[1 rows x 15 columns]

**Test**

In [ ]:
patients_copy.describe()

patient_id     weight      height         bmi
count  503.000000  503.00000  503.000000  503.000000
mean   252.000000  173.55186   66.634195   27.483897
std    145.347859   33.58609    4.411297    5.276438
min      1.000000  102.10000   27.000000   17.100000
25%    126.500000  149.30000   63.000000   23.300000
50%    252.000000  175.30000   67.000000   27.200000
75%    377.500000  199.50000   70.000000   31.750000
max    503.000000  255.90000   79.000000   37.700000

**Define**

Replacing the outlier of height by replace function.

**Code**

In [ ]:
height_inches=patients_copy['height'].replace(27,72)

In [ ]:
height_inches

0      66
1      66
2      71
3      70
4      72
       ..
498    72
499    70
500    67
501    67
502    71
Name: height, Length: 503, dtype: int64

In [ ]:
patients_copy.loc[:,'height']=height_inches.values

**Test**

In [ ]:
patients_copy.describe()

patient_id     weight      height         bmi
count  503.000000  503.00000  503.000000  503.000000
mean   252.000000  173.55186   66.723658   27.483897
std    145.347859   33.58609    4.047179    5.276438
min      1.000000  102.10000   59.000000   17.100000
25%    126.500000  149.30000   63.000000   23.300000
50%    252.000000  175.30000   67.000000   27.200000
75%    377.500000  199.50000   70.000000   31.750000
max    503.000000  255.90000   79.000000   37.700000

**Define**

The bmi will be affected as the weight and height has been corrected.

**Code**

In [ ]:
patients_copy['bmi']=703*patients_copy['weight']/(patients_copy['height']*patients_copy['height'])

In [ ]:
patients_copy['bmi']=patients_copy['bmi'].round(1)

**Test**

In [ ]:
patients_copy.describe()

patient_id     weight      height         bmi
count  503.000000  503.00000  503.000000  503.000000
mean   252.000000  173.55186   66.723658   27.483897
std    145.347859   33.58609    4.047179    5.276438
min      1.000000  102.10000   59.000000   17.100000
25%    126.500000  149.30000   63.000000   23.300000
50%    252.000000  175.30000   67.000000   27.200000
75%    377.500000  199.50000   70.000000   31.750000
max    503.000000  255.90000   79.000000   37.700000

**Define**

Reducing some memory usage.

**Code**

In [ ]:
patients_copy['patient_id']=patients_copy['patient_id'].astype('int32')

In [ ]:
patients_copy['country']=patients_copy['country'].astype('category')

In [ ]:
patients_copy['height']=patients_copy['height'].astype('int32')

**Test**

In [ ]:
patients_copy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 15 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   patient_id    503 non-null    int32         
 1   assigned_sex  503 non-null    category      
 2   given_name    503 non-null    object        
 3   surname       503 non-null    object        
 4   address       491 non-null    object        
 5   city          491 non-null    object        
 6   state         491 non-null    object        
 7   zip_code      491 non-null    object        
 8   country       491 non-null    category      
 9   birthdate     503 non-null    datetime64[ns]
 10  weight        503 non-null    float64       
 11  height        503 non-null    int32         
 12  bmi           503 non-null    float64       
 13  phone_number  491 non-null    object        
 14  email         491 non-null    object        
dtypes: category(2), datetime64[ns](1), float

In [ ]:
treatment_copy.sample(5)

given_name    surname  hba1c_start  ...  start   end adverse_reaction
64      johana       hrdá         7.79  ...   54u    54u              NaN
131       kate  wilkinson         7.72  ...   36u    39u              NaN
80       annie      allen         9.95  ...   36u    42u              NaN
53     burrell    mathieu         8.71  ...   46u    58u              NaN
306     csilla   herczegh         7.71  ...   43u    46u              NaN

[5 rows x 9 columns]

In [ ]:
treatment_copy.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 350 entries, 0 to 349
Data columns (total 9 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   given_name        350 non-null    object 
 1   surname           350 non-null    object 
 2   hba1c_start       350 non-null    float64
 3   hba1c_end         350 non-null    float64
 4   hba1c_change      350 non-null    float64
 5   treatment         350 non-null    object 
 6   start             350 non-null    object 
 7   end               350 non-null    object 
 8   adverse_reaction  35 non-null     object 
dtypes: float64(3), object(6)
memory usage: 27.3+ KB


**Define**

Removing one duplicate row in treatment dataframe with the help of drop_duplicates().

**Code**

In [ ]:
treatment_copy.duplicated().sum()

1

In [ ]:
treatment_copy=treatment_copy.drop_duplicates()

**Test**

In [ ]:
treatment_copy.shape

(349, 9)

**Define**

Making the first letter of the given_name and surname in capital letter using str.capitalize().

**Code**

In [ ]:
treatment_copy['given_name']=treatment_copy['given_name'].str.capitalize()

In [ ]:
treatment_copy['surname']=treatment_copy['surname'].str.capitalize()

**Test**

In [ ]:
treatment_copy.head()

given_name     surname  hba1c_start  ...  start   end adverse_reaction
0   Veronika    Jindrová         7.63  ...   41u    48u              NaN
1       Skye  Gormanston         7.97  ...   33u    36u              NaN
2     Sophia      Haugen         7.65  ...   37u    42u              NaN
3      Eddie      Archer         7.89  ...   31u    38u              NaN
4       Asia     Woźniak         7.76  ...   30u    36u              NaN

[5 rows x 9 columns]

**Define**

Removing the given_name and surname from the treatment column and merging the patient_id column with the help of merge function.

**Code**

In [ ]:
id_names=patients_copy[['patient_id','given_name','surname']]

In [ ]:
id_names

patient_id given_name      surname
0             1        Zoe      Wellish
1             2     Pamela         Hill
2             3        Jae       Debord
3             4       Liêm         Phan
4             5        Tim      Neudorf
..          ...        ...          ...
498         499    Mustafa    Lindström
499         500      Ruman      Bisliev
500         501      Jinke    de Keizer
501         502    Chidalu  Onyekaozulu
502         503        Pat      Gersten

[503 rows x 3 columns]

In [ ]:
treatment_copy=pd.merge(treatment_copy, id_names, on=['given_name', 'surname'])

In [ ]:
treatment_copy = treatment_copy.drop(['given_name', 'surname'], axis=1)

**Test**

In [ ]:
treatment_copy.head()

hba1c_start  hba1c_end  hba1c_change  ...   end adverse_reaction patient_id
0         7.63       7.20          0.43  ...   48u              NaN        225
1         7.97       7.62          0.35  ...   36u              NaN        242
2         7.65       7.27          0.38  ...   42u              NaN        345
3         7.89       7.55          0.34  ...   38u              NaN        276
4         7.76       7.37          0.39  ...   36u              NaN         15

[5 rows x 8 columns]

**Define**

Removing all the John Doe records from the patients_copy dataframe.

**Code**

In [ ]:
patients_copy=patients_copy[patients_copy['surname']!='Doe']

**Test**

In [ ]:
patients_copy.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 497 entries, 0 to 502
Data columns (total 15 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   patient_id    497 non-null    int32         
 1   assigned_sex  497 non-null    category      
 2   given_name    497 non-null    object        
 3   surname       497 non-null    object        
 4   address       485 non-null    object        
 5   city          485 non-null    object        
 6   state         485 non-null    object        
 7   zip_code      485 non-null    object        
 8   country       485 non-null    category      
 9   birthdate     497 non-null    datetime64[ns]
 10  weight        497 non-null    float64       
 11  height        497 non-null    int32         
 12  bmi           497 non-null    float64       
 13  phone_number  485 non-null    object        
 14  email         485 non-null    object        
dtypes: category(2), datetime64[ns](1), float

In [ ]:
patients_copy.sample(5)

patient_id assigned_sex  ...       phone_number                         email
291         292         male  ...       302-698-2057      EncoZibrik@superrito.com
110         111         male  ...  +1 (214) 637-0855  MiroslavStepanek@teleworm.us
311         312       female  ...       601-389-7682    EricaMacDonald@teleworm.us
453         454         male  ...       281-556-4376              WuSung@gustr.com
240         241       female  ...  +1 (256) 615-5522    MarphisaCompagnon@cuvox.de

[5 rows x 15 columns]

In [ ]:
treatment_copy.sample(5)

hba1c_start  hba1c_end  hba1c_change  ...   end adverse_reaction patient_id
31          7.63       7.30          0.33  ...   46u              NaN        160
89          7.82       7.28          0.54  ...   50u              NaN        361
98          7.99       7.72          0.27  ...   51u              NaN         66
190         7.62       7.16          0.46  ...   30u              NaN        185
109         7.64       7.31          0.33  ...   61u     hypoglycemia         14

[5 rows x 8 columns]

**Define**

Removing the same person with nicknames by drop method.

**Code**

In [ ]:
patients_copy['surname'].value_counts()

Jakobsen       3
Taylor         3
Batukayev      2
Johnson        2
Lương          2
              ..
Maslov         1
Teichelmann    1
Nakahara       1
Glockner       1
Bisliev        1
Name: surname, Length: 465, dtype: int64

In [ ]:
patients_copy[patients_copy.surname == 'Taylor']

patient_id assigned_sex  ...  phone_number                       email
131         132       female  ...  304-438-2648    SandraCTaylor@dayrep.com
282         283       female  ...  304-438-2648    SandraCTaylor@dayrep.com
426         427         male  ...  305-434-6299  RogelioJTaylor@teleworm.us

[3 rows x 15 columns]

In [ ]:
patients_copy[patients_copy.surname == 'Jakobsen']

patient_id assigned_sex  ...       phone_number                         email
24           25         male  ...  +1 (845) 858-7707     JakobCJakobsen@einrot.com
29           30         male  ...  +1 (845) 858-7707     JakobCJakobsen@einrot.com
432         433       female  ...       979 203 0438  KarenJakobsen@jourrapide.com

[3 rows x 15 columns]

In [ ]:
patients_copy[patients_copy.surname == 'Gersten']

patient_id assigned_sex  ...  phone_number                     email
97           98         male  ...  402-848-4923  PatrickGersten@rhyta.com
502         503         male  ...  402-848-4923  PatrickGersten@rhyta.com

[2 rows x 15 columns]

In [ ]:
patients_copy=patients_copy.drop([282, 29, 502])

**Test**

In [ ]:
patients_copy['surname'].value_counts()

Collins        2
Grímsdóttir    2
Nilsen         2
Aranda         2
Silva          2
              ..
Maslov         1
Teichelmann    1
Nakahara       1
Glockner       1
Bisliev        1
Name: surname, Length: 465, dtype: int64

**Define**

Filling the NaN values of the columns with a proper statement and the NaN values for the country as 'United States'.

**Code**

In [ ]:
patients_copy['email']=patients_copy['email'].fillna('Email not found')

In [ ]:
patients_copy['address']=patients_copy['address'].fillna('Address not found')

In [ ]:
patients_copy['city']=patients_copy['city'].fillna('City not found')

In [ ]:
patients_copy['state']=patients_copy['state'].fillna('State not found')

In [ ]:
patients_copy['zip_code']=patients_copy['zip_code'].fillna('Zip code not found')

In [ ]:
patients_copy['country']=patients_copy['country'].fillna('United States')

In [ ]:
patients_copy['phone_number']=patients_copy['phone_number'].fillna('Phone number not found')

In [ ]:
treatment_copy['adverse_reaction']=treatment_copy['adverse_reaction'].fillna('Information Not Available')

**Test**

In [ ]:
patients_copy.sample(5)

patient_id  ...                              email
70           71  ...  AlexanderEMathiesen@superrito.com
241         242  ...          SkyeGormanston@dayrep.com
358         359  ...      AmalieJChristensen@einrot.com
336         337  ...      SharonLCollins@jourrapide.com
257         258  ...                    Email not found

[5 rows x 15 columns]

In [ ]:
treatment_copy.sample(5)

hba1c_start  hba1c_end  ...           adverse_reaction patient_id
120         9.48       9.06  ...  Information Not Available         53
219         7.88       7.50  ...                      cough        259
234         7.50       7.08  ...  Information Not Available        271
222         7.73       7.32  ...  Information Not Available        123
326         7.68       7.36  ...  Information Not Available        262

[5 rows x 8 columns]